hamilflow/models/pendulum.py

```diff
@@ -0,0 +1,146 @@
+import math
+from functools import cached_property
+from typing import Any, Dict, Optional, Union
+
+import numpy as np
+import pandas as pd
+from numpy.typing import ArrayLike
+from pydantic import BaseModel, Field, computed_field
+from scipy.special import ellipj, ellipk
+
+
+class PendulumSystem(BaseModel):
+    r"""The params for the pendulum
+
+    :param float omega0: $\omega_0 \coloneqq \sqrt{\frac{U}{I}} > 0$, frequency
+    parameter
+    """
+
+    omega0: float = Field(gt=0)
+
+    def __init__(self, omega0: float, **kwargs: Any) -> None:
+        return super().__init__(omega0=omega0, **kwargs)
+
+
+class PendulumIC(BaseModel):
+    r"""The initial condition for a pendulum
+
+    :param float theta0: $-\frac{\pi}{2} \le \theta_0 \le \frac{\pi}{2}$, the
+    initial angle
+    """
+
+    theta0: float = Field(ge=-math.pi / 2, le=math.pi / 2)
+
+    def __init__(self, theta0: float, **kwargs: Any) -> None:
+        return super().__init__(theta0=theta0, **kwargs)
+
+    @computed_field  # type: ignore[misc]
+    @cached_property
+    def k(self) -> float:
+        r"""A convenient number
+
+        :return float: $\sin\frac{\theta_0}{2}$
+        """
+        return math.sin(self.theta0 / 2)
+
+
+class Pendulum:
+    r"""Generate time series data for a pendulum.
+
+    We describe a generic pendulum system by the Lagrangian action
+    $$
+    S_L\[\theta\] = I \int_{t_0}^{t_1} \mathbb{d}t
+    \left\\{\frac{1}{2} \dot\theta^2 + \omega_0^2 \cos\theta \right\\}\,,
+    $$
+    where $\theta$ is the _angle_ from the vertical to the pendulum;
+    $I$ is the _inertia parameter_ introduced for dimensional reasons,
+    and $\omega_0$ the _frequency parameter_.
+
+    Details are collected in the tutorial.
+    """
+
+    def __init__(
+        self,
+        system: Union[int, float, Dict[str, Union[int, float]]],
+        initial_condition: Union[int, float, Dict[str, Union[int, float]]],
+    ) -> None:
```

Comment:
Init always returns `None` so it may not be necessary to type hint the returns of `__init__`. But I am also okay with keeping it here. No strong opinion on this.